taskmanager/server.go

```diff
@@ -0,0 +1,155 @@
+	Register(ctx context.Context, t *task.Task) (string, error)
+	Lookup(id string) *task.Task
+}
+
+// Server is an implementation of the TaskManager API
+type Server struct {
+	api.UnimplementedTaskManagerServer
+	registry TaskRegistry
+}
+
+// New creates and initialises a new Server with default settings
+func New() *Server {
+	return withRegistry(registry.New())
+}
+
+// withRegistry creates and initialises a new Server with a custom task registry
+// implementation
+func withRegistry(registry TaskRegistry) *Server {
+	return &Server{registry: registry}
+}
+
+// StartTask attempts to start and register a task with the task manager.
+func (server *Server) StartTask(ctx context.Context, req *api.StartTaskRequest) (*api.StartTaskResponse, error) {
+	t := task.New(
+		req.GetBinary(),
+		req.GetWorkingDir(),
+		req.GetEnvironment(),
+		req.GetArgs()...)
+
+	// Start the task *before* registering it, as we don't want to expose the task
+	// to RPC calls from the outside world until the task is in a state that can
+	// handle it.
+	err := t.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	// Register the task with the registry, which will make it available to other
+	// RPC calls
+	id, err := server.registry.Register(ctx, t)
+	if err != nil {
+		killTask(t) // Don't leak the task if we fail to register it
+		return nil, err
+	}
+
+	// Give the caller a handle to their task
+	return &api.StartTaskResponse{
+		TaskId: &api.TaskHandle{Id: id},
+	}, nil
+}
+
+// FetchLogs extracts and returns the collected stdout & stderr data from the
+// task
+func (server *Server) FetchLogs(
+	ctx context.Context, req *api.FetchLogsRequest) (*api.FetchLogsResponse, error) {
+
+	task := server.registry.Lookup(req.TaskId.Id)
+	if task == nil {
+		return nil, NoSuchTask
+	}
+
+	response := &api.FetchLogsResponse{
+		Stdout: task.Stdout(),
+		Stderr: task.Stderr(),
+	}
+
+	return response, nil
+}
+
+// QueryTask fetches information about a given task
+func (server *Server) QueryTask(
+	ctx context.Context, req *api.QueryTaskRequest) (*api.QueryTaskResponse, error) {
+
+	task := server.registry.Lookup(req.TaskId.Id)
+	if task == nil {
+		return nil, NoSuchTask
+	}
+
+	var exitCode *int32
+	status, taskExitCode := task.Status()
+	if taskExitCode != -1 {
+		exitCode = new(int32)
+		(*exitCode) = int32(taskExitCode)
+	}
+
+	response := &api.QueryTaskResponse{
+		StatusCode: status,
+		ExitCode:   exitCode,
+	}
+
+	return response, nil
+}
+
+// Signal requests that a task should be stopped
+func (server *Server) Signal(
+	ctx context.Context, req *api.SignalTaskRequest) error {
+
+	task := server.registry.Lookup(req.TaskId.Id)
+	if task == nil {
+		return NoSuchTask
+	}
+
+	// Note the hardcoded 5s timeout here; this shoud at the very least be
+	// a be parameter of the Server, preferably configurable somehow by the
+	// user. For the sake of this exercise, it's just a hardcoded value.
+	signalCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

Comment:
Any reason the context passed to this server method isn't used as a parent here?

Reply:
Because the signal timeout (which is implemented as waiting on a `context.Done()`) is not bounded by the lifetime of this function call.

Once this call sends the signal its done, and will return to the caller. The caller may then (and in fact, probably _will_) cancel the enclosing context. If the signal context was a child of the enclosing context, this would end up killing the task early.